nifi-registry-core/nifi-registry-security-api/src/main/java/org/apache/nifi/registry/security/authentication/AuthenticationRequest.java

```diff
@@ -17,17 +17,20 @@
 package org.apache.nifi.registry.security.authentication;
 
 import java.io.Serializable;
+import java.util.Objects;
 
 public class AuthenticationRequest implements Serializable {
 
     private String username;
     private Object credentials;
     private Object details;
+    private String httpMethod;
 
-    public AuthenticationRequest(String username, Object credentials, Object details) {
+    public AuthenticationRequest(String username, Object credentials, Object details, String httpMethod) {
```

Comment:
This works, but I'd prefer not to add the HttpMethod info to all instances of AuthenticationRequest as it is only every really used by the X509IdentityProvider.

The `Object details` argument of this constructor is designed to be a generic way that additional details relevant to the AuthenticationRequest can be passed from the `IdentityProvider.extractCredentials` method to the IdentityProvider implementation later for authentication. `details` can be any Object type, it is only later read by the same `IdentityProvider` instance or the AuthenticationProvider for that same IdentityProvider. In most cases, it is unused.

Currently, for the `X509IdentityProvider`, details is set to the String value of the Proxied Entities Chain Header, as that was the only relevant "detail" at the time. I would suggest just creating a new wrapper object such as X509AuthenticationRequestDetails that can hold the proxied entities chain and the HTTP method. That can be created and set in `X509IdentityProvider.extractCredentials`, and later both values can be extracted in `X509IdentityAuthenticationProvider.buildAuthenticatedToken`,

This is a small nitpick, but I'd prefer not to change AuthenticationRequest if necessary as it is part of the extension point API for third-parties, so changing this could break and third-party impls out there. I know it would also be possible to keep the current 3-arg constructor and add a 4-arg constructor, but that looses some of the benefits of this implementation such as ensuring non-null, and also adds complexity to an extension point that most people won't care about for authentication purposes. Encasing the http method in details prevents any other IdentityProviders (ours or third-parties) from needing updating, but still lets us do our authorization-as-part-of-authentication check for proxies.

Reply:
Makes sense, now that I understand how the details object is being used I definitely agree that would be the better approach. Will work on updating that. Thanks!